File: Scripts/Librairies/jonswap.py
"""
Informations
============

    * Fichier : ``Knu/Scripts/Librairies/jonswap.py``
    * Auteur: Victor Baconnet
    * Date de dernière modification: 27 août 2021

Description
===========

Fonctions pour générer un spectre JONSWAP et les coefficients d'amplitude, périodes,
et déphasages pour générer le signal de houle irrégulière. 

Génération du spectre
----------------------

Le spectre de JONSWAP est généré selon la formule suivante

.. math::
    
    S(\omega,a) = aH_s^2 \omega_p^4 \omega^{-5} e^{-\\frac{3}{4}\\left( 
        \\frac{\omega_p}{\omega}\\right)^4} \gamma^b

avec 

.. math::
    
    b=\exp{-\\frac{(\omega - \omega_p)^2}{2\sigma^2\omega_p^2}}
   
et :math:`\sigma=0.07` si :math:`\omega < \omega_p`, :math:`\sigma=0.09` sinon. La constante 
réelle :math:`a` peut être déterminée en utilisant la définition de :math:`H_s`

.. math::
    H_s = 4\sqrt{\int_{0}^{+\infty}{S(\omega,a)d\omega}}


De telle sorte que :

.. math::
    a = \\frac{H_s^2}{16}\\left(\int_{0}^{+\infty}{S(\omega, a=1)d\omega}\\right)^{-1}


Génération des amplitudes
--------------------------

Les hauteurs crête-à-creux sont générées avec la relation:
    
.. math::
    \\frac{H_i}{2} = \sqrt{S_i\Delta\omega}

Génération des déphasages
--------------------------

Les déphasages sont générés aléatiroement entre :math:`0` et :math:`2\pi`.

Utilisation
===========

Les fonctions contenues dans ce fichier sont à réutiliser dans d'autres fichiers 
pour générer de la houle irrégulière, comme dans ``genHouleIrreguliere``.

L'éxecution de ce fichier n'affichera aucun résultat.

Exemple d'utilisation
======================

.. code-block:: python
   
    import numpy as np
    import matplotlib.pyplot as plt    
    from jonswap import genJonswapParams, jonswap
    
    # Paramètre du spectre jonswap
    Tmin = 0.87
    Tmax = 2.19
    Tp = 1.57
    Hs = 0.125
    gamma = 3.3
    
    wmin = 2.0*np.pi/Tmax #Fréquence min
    wmax = 2.0*np.pi/Tmin #Fréquence max
    wp   = 2.0*np.pi/Tp   #Pulsation pic
    Ncomposantes = 1000
    dw = (wmax - wmin)/Ncomposantes #Incrément de fréquence
    w = np.arange(wmin,wmax+dw,dw)  #Génération du vecteur de fréquences
    
    # Génération des Amplitudes, déphasages et spectre
    spectre = jonswap(Hs,Tp,gamma,w)
    H, phi = genJonswapParams(spectre, w)
    
    # Génération de houle irrégulière
    t = np.linspace(0.0,100,1001)
    
    houleIrreg = 0.0
    for i in range(len(H)):
        houleIrreg += 0.5*H[i]*np.sin(w[i]*t + phi[i])
    
    # Affichage des résultats
    fig, axs = plt.subplots(figsize = (10,8), nrows = 2)
    
    axs[0].plot(w, spectre)
    axs[0].set_xlabel(r"$\omega$ (rad/s)")
    axs[0].set_ylabel(r"$S(\omega)$ (m²s)")
    axs[0].set_xlim((wmin, wmax))
    axs[0].grid()
    axs[0].set_title(rf"Spectre de paramètres Hs = {Hs}, Tp = {Tp} et $\gamma$ = {gamma}")
    
    axs[1].plot(t,houleIrreg)
    axs[1].set_xlabel("Temps (s)")
    axs[1].set_ylabel(r"$\eta(t)$ (m)")
    axs[1].set_xlim((0, 100))
    axs[1].grid()
    axs[1].set_title("Signal de houle irrégulière")

Dépendances
===========

* ``numpy``
* ``matplotlib``
* ``sys``, 

Code source
============

Lien vers le `code source <https://github.com/victor13165/Knu/blob/main/Scripts/Librairies/jonswap.py>`_

Fonctions
=========
"""

# ---------  module indispensable ----------------
import numpy as np
# ------------------------------------------------

def integre(vec,dx):
    """Intègre un vecteur selon la méthode des trapèzes 
    
    :param vec: Vecteur de valeurs à intégrer
    :type vec: ``numpy.ndarray`` ou itérable
    :param dx: Pas de discrétisation
    :type dx: ``float``
    :return: res : scalaire correspondant à l'intégrale du vecteur
    :rtype: ``float``
    """
    res = 0
    
    for i in range(len(vec)-1):
        res += (vec[i+1]+vec[i])*0.5
        
    return res*dx

def sigma(w,wp):
    """Calcule le coefficient :math:`\sigma = 0.07` si :math:`\omega < \omega_p` et 
    :math:`\sigma = 0.09` si :math:`\omega > \omega_p`, à partir d'une combinaison de
    fonctions de Heaviside. Cela permet de ne pas écrire avec des "if" pour 
    gagner du temps si on utilise un vecteur w très grand.

    :param w: Valeur ou vecteur de fréquences en rad/s
    :type w: ``float`` ou ``numpy.ndarray``
    :param wp: fréquence pic en rad/s
    :type  wp:  ``float``
    :return: 0.07 si :math:`\omega < \omega_p`, 0.09 sinon
    :rtype: ``float``
    """

    termeInferieur = 0.07 * np.heaviside(wp-w, 0.07);
    termeSuperieur = 0.09 * np.heaviside(w-wp, 0.00);
    return termeInferieur + termeSuperieur

def spectreJonswap(a, Hs, Tp, gamma, w):
    """Calcule le spectre de Jonswap avec la formule donnée dans le manuel du canal à houle

    :param a: le coefficient qui doit faire l'objet de l'optimisation
    :type a: ``float``
    :param Hs: hauteur caractéristique
    :type Hs: ``float``
    :param Tp: période pic
    :type Tp: ``float``
    :param gamma: Coefficient d'élancement
    :type gamma: ``float``
    :param w: vecteur de fréquences en rad/s
    :type w: ``numpy.ndarray``
    :return: vecteur densité spectrale de puissance de la même taille que w
    """
    wp = 2.0*np.pi/Tp
    t1 = a * Hs**2 * wp**4 * w**(-5)
    t2 = np.exp( -1.25*(w/wp)**(-4) )
    exposant = np.exp( -0.5 * (w-wp)**2 / (sigma(w,wp)**2 * wp**2) )
    t3 = gamma**exposant
    return t1*t2*t3


def jonswap(Hs, Tp, gamma, w):
    """Renvoie une densité spectrale de puissance selon le modèle JONSWAP, avec 
    le calcul de la constante a
    
    :param Hs: Hauteur caractéristique
    :type Hs: ``float``
    :param Tp: Période pic
    :type Tp: ``float`` 
    :param gamma: Coefficient d'élancement
    :type gamma: ``float`` 
    :param w: Fréquence en rad/s
    :type w: ``float`` ou ``numpy.ndarray``
    :return: Densité spectrale de puissance selon le modèle JONSWAP
    :rtype: ``float`` ou ``numpy.ndarray``
    """
    
    assert Tp != 0

    spectre = spectreJonswap(1.0, Hs, Tp, gamma, w) #Calcul du spectre avec a=1    
    dw = w[1] - w[0]
    a = ( Hs / (4.0 * np.sqrt(integre(spectre,dw))) )**2 
    return spectreJonswap(a, Hs, Tp, gamma, w) #Calcul du spectre avec cette valeur de a


def genJonswapParams(spectreJonswap, w):
    """
    Calcule les amplitudes, déphasages et spectre pour générer une onde irrégulière

    :param spectreJonswap: le spectre de densité d'énergie JONSWAP
    :type spectreJonswap: ``float``
    :param w: Fréquence en rad/s
    :type w: ``float`` ou ``numpy.ndarray``
    
    :return: Amplitudes et déphasages (aléatoires entre :math:`0` et :math:`2\pi`)
    :rtype: ``float`` ou ``numpy.ndarray``
    """
    
    dw = w[1] - w[0]
    
    H = 2.0 * np.sqrt(2.0 * spectreJonswap * dw) #Génération d'amplitudes
    dephasages = 2.0*np.pi*np.random.rand(len(w))   #Génération de déphasage entre 0 et 2pi
    
    return H, dephasages

File: Scripts/Canal/houle_irreguliere/jonswap.m
% Génération spectre de Jonswap

% Calcul du spectre de jonswap en fonction de la fréquence et des
% paramètres Hs (Hm0), Tp et gamma

% Note: Il y a deux méthodes pour obtenir le spectre
%       --> Formule avec ajustement du coefficient a par intégration
%       --> Formule dans alea.f
% Pour choisir celle à utiliser, ajuster la variable "type". Si type est
% différent de 1, la méthode sélectionnée par défaut est l'ajustement de a.
% Si type = 1, on sélectionne la méthode de alea.f
function spectre = jonswap(w, Hs, Tp, gamma, type)

    if type ~= 1
        %Calcul du spectre avec a=1 
        spectre_1 = spectreJonswap(1.0, w, Hs, Tp, gamma); 
        
        dw = w(2) - w(1); % Pas de discrétisation
        a = (Hs / 4.0)^2 * (integrale(spectre_1, dw))^(-1); % Calcul de a
        
        % Spectre final
        spectre = spectreJonswap(a, w, Hs, Tp, gamma);
    else
        f = w / (2.0*pi);
        wp = 2.0*pi/Tp;
        fp = wp / (2.0*pi);
        
        betaJ = (0.06238/(0.23+0.0336*gamma-0.185/(1.9+gamma)))* ...
            (1.094-0.01915*log(gamma));
        %disp('betaJ : '  +  betaJ)
        
        expo = exp(-1.25 * ((Tp*f).^(-4.0)) );
        
        gexpo = gamma .^ (exp( -0.5 * (f./fp - 1.0).^2 / (sig(w,wp)^2) ));
        
        spectre = betaJ * Hs * Hs * Tp^(-4.0) * f.^(-5.0) .* expo .* gexpo;
    end
    
end

% Spectre de Jonswap en fonction de la de la fréquence et des
% paramètres Hs (Hm0), Tp et gamma, ainsi que la constante a
function spec = spectreJonswap(a, w, Hs, Tp, gamma)
    wp = 2.0*pi/Tp;  % Pulsation pic
    
    t1 = 2.0*pi * a * Hs^2 * wp^(4) * w.^(-5);     % terme 1
    t2 = exp( -1.25*((w./wp).^(-4)) );    % terme 2
    
    exposant = exp( -0.5 * (w-wp).^2 / (sig(w, wp)^2 * wp^2) );
    t3 = gamma.^exposant;               % terme 3
    
    spec = t1.*t2.*t3;  % Spectre final
end

% Intègre un vecteur avec un pas constant par la méthode des trapèzes
function res = integrale(vec, dx)
    N = length(vec);
    res = sum(vec(2:N) + vec(1:N-1)) * 0.5 * dx;
end

% Renvoie la valeur de sigma selon omega et omega_p
function s = sig(w, wp)
    if w < wp
        s = 0.07;
    else
        s = 0.09;
    end
end